### PR TITLE
Mobile: adds redo undo functionality in mobile

### PR DESCRIPTION
**Pull request by amitsin6h**
**Android Joplin app**
- This PR adds undo/redo functionality on the app
- Created UndoRedoService.ts which has undo/redo logic 
- Improved the code quality, also proper comments have been made
- Reference of issue #2595   

> POC of undo/redo functionality

WhatsApp Video 2020-03-15 at 6 27 01 PM

### CODE REVIEW

**Comment by amitsin6h** [comment]
I am working to fix Travis build, I know why it failed

**Comment by sealch** [comment]
What is the point of this behaviour with text? In desktop Joplin app if you press Ctrl/Cmd + Z it erases the whole word, not letter by letter

Screen Recording 2020-03-22 at 19 15 22

**Comment by amitsin6h** [comment]
@sealch  I know but in the mobile app, this is the best way I can do. If you have some better approach or ideas in handling undo/redo events on mobile. Please do let me know I'll be happy to improve my code.

**Comment by amitsin6h** [comment]
@PackElend please label me.

File: ReactNativeClient/lib/services/UndoRedoService.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+		this.textHistory = [];
+		this.text = '';
+	}
+
+	// It will, increment totalKeyEventCount on text change and get the text
+	// to store with refrence to totalKeyEventCount inside textHistory[]
+	handleText(text:string) {
+		this.totalKeyEventCount++;
+		this.textHistory[this.totalKeyEventCount] = text;
+	}
```
Comment (laurent22) [inline ReactNativeClient/lib/services/UndoRedoService.ts]: This class is hard to make sense of and it's good you've added comments, however comments should explain **why** the code does something, **not what it does**.

Literally all your comments repeat what the code does. I can read code, I know ++ means "increment", and ">" means "greater than", but I don't know why your class needs to keep track of event count, why array elements are added directly by index, and not with "push", why elements are not discarded after undo, why you need to store a "text" value when the undo stack should give already give you that value, etc.

totalKeyEventCount and totalClickEventCount are of "any" type, but they must be numbers based on your code.

I'm sorry I'm not very keen on the design of this class.

**Comment by laurent22** [comment]
The UndoRedoService is hard to make sense of and it's good you've added comments, however comments should explain **why** the code does something, **not what it does**.

Literally all your comments repeat what the code does. I can read code, I know ++ means "increment", and ">" means "greater than", but I don't know why your class needs to keep track of event count, why array elements are added directly by index, and not with "push", why elements are not discarded after undo, why you need to store a "text" value when the undo stack should give already give you that value, etc.

totalKeyEventCount and totalClickEventCount are of "any" type, but they must be numbers based on your code.

I'm sorry maybe it's just me but I'm not very keen on the design of this class.

**Comment by amitsin6h** [comment]
Okay, I will improve the design of UndoRedoServices class

**Comment by laurent22** [comment]
If you want to give it another try, perhaps you could check what are the standard ways to implement an undo/redo stack. For example have a look at some generic algorithms for this, even start with Wikipedia - https://en.wikipedia.org/wiki/Undo#Undo_implementation

Basically don't start by looking for existing JS scripts, start by understanding the best way to implement it. Then you might be able to find an existing JS script but you'll know what to look for and you'll be able to decide if a script is good or not.

**Comment by laurent22** [comment]
Closing for now. Please resubmit once it's ready.

**Pull request closed without merging**